CMakeLists.txt

```diff
@@ -485,6 +489,8 @@ endif()
 
 unset(CMAKE_REQUIRED_FLAGS)
 
+set(TS_USE_HELLO_CB ${HAVE_SSL_CTX_SET_CLIENT_HELLO_CB} | ${HAVE_SSL_CTX_SET_SELECT_CERTIFICATE_CB})
```

Comment:
Shouldn't it be logical or ( || ) not bitwise or?

Reply:
I had an even more suspicious reaction than this when I saw it, so I tested locally. My intuition was spot on:

```cmake
set(MYVAR TRUE | FALSE)
message(STATUS "Value of MYVAR is ${MYVAR}.")
```

```
% cmake -P myfile.cmake
-- Value of MYVAR is TRUE;|;FALSE.
```

Reply:
I think what's wanted here is a generator expression, something like
```cmake
"<$OR:<$BOOL:HAVE_SSL_CTX_SET_CLIENT_HELLO_CB> <$BOOL:HAVE_SSL_SET_SELECT_CERTIFICATE_CB>>"
```

or use an `if/else`.

Reply:
Or just do the **OR** in C++:
```
$ git diff
diff --git a/include/tscore/ink_config.h.cmake.in b/include/tscore/ink_config.h.cmake.in
index 260ea7a24..bd4e82a47 100644
--- a/include/tscore/ink_config.h.cmake.in
+++ b/include/tscore/ink_config.h.cmake.in
@@ -143,7 +143,6 @@ const int DEFAULT_STACKSIZE = @DEFAULT_STACK_SIZE@;
 #cmakedefine01 TS_USE_FAST_SDK
 #cmakedefine01 TS_ENABLE_FIPS
 #cmakedefine01 TS_USE_GET_DH_2048_256
-#cmakedefine01 TS_USE_HELLO_CB
 #cmakedefine01 TS_USE_HWLOC
 #cmakedefine01 TS_USE_KQUEUE
 #cmakedefine01 TS_USE_LINUX_IO_URING
@@ -156,6 +155,8 @@ const int DEFAULT_STACKSIZE = @DEFAULT_STACK_SIZE@;
 #cmakedefine01 TS_USE_TLS_ASYNC
 #cmakedefine01 TS_USE_TPROXY
 
+#define TS_USE_HELLO_CB (HAVE_SSL_CTX_SET_CLIENT_HELLO_CB || HAVE_SSL_CTX_SET_SELECT_CERTIFICATE_CB)
+
 #cmakedefine01 TS_HAS_VERIFY_CERT_STORE
 #cmakedefine01 TS_HAS_TLS_SESSION_TICKET
 #cmakedefine01 TS_HAS_TLS_EARLY_DATA
$ 
```

Reply:
I copied what we do for `TS_HAS_TLS_EARLY_DATA`. That one needs to be fixed as well if this doesn't work.

Reply:
When I make these changes:
```
$ git diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5cbf1d6fd..1b004ab41 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -495,6 +495,12 @@ set(TS_HAS_TLS_EARLY_DATA
     ${HAVE_SSL_IN_EARLY_DATA}
 )
 
+set(TS_DUMMY
+    ${HAVE_BIO_GET_EX_DATA}
+    |
+    ${HAVE_BIO_SET_EX_DATA}
+)
+
 set(CMAKE_REQUIRED_INCLUDES netinet/in.h netinet/tcp.h)
 check_type_size("struct tcp_info" STRUCT_TCP_INFO)
 unset(CMAKE_REQUIRED_INCLUDES)
diff --git a/include/tscore/ink_config.h.cmake.in b/include/tscore/ink_config.h.cmake.in
index fe85c656c..f8b0c4dad 100644
--- a/include/tscore/ink_config.h.cmake.in
+++ b/include/tscore/ink_config.h.cmake.in
@@ -83,6 +83,7 @@
 #cmakedefine HAVE_BIO_GET_EX_NEW_INDEX 1
 #cmakedefine HAVE_BIO_GET_EX_DATA 1
 #cmakedefine HAVE_BIO_SET_EX_DATA 1
+#cmakedefine01 TS_DUMMY
 #cmakedefine HAVE_BIO_METH_GET_CTRL 1
 #cmakedefine HAVE_BIO_METH_GET_CREATE 1
 #cmakedefine HAVE_BIO_METH_GET_DESTROY 1
$ 
```
This is the relevant portion of the generated ink_config.h:
```
/* #undef HAVE_BIO_GET_EX_DATA */
/* #undef HAVE_BIO_SET_EX_DATA */
#define TS_DUMMY 1
```
I think the cmakedefine directives check if the variable is empty or not.

Reply:
This seems to work:
```
$ git diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5cbf1d6fd..71c2eba2f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -495,6 +495,9 @@ set(TS_HAS_TLS_EARLY_DATA
     ${HAVE_SSL_IN_EARLY_DATA}
 )
 
+set(TS_DUMMY ${HAVE_BIO_GET_EX_DATA}${HAVE_BIO_SET_EX_DATA}
+)
+
 set(CMAKE_REQUIRED_INCLUDES netinet/in.h netinet/tcp.h)
 check_type_size("struct tcp_info" STRUCT_TCP_INFO)
 unset(CMAKE_REQUIRED_INCLUDES)
diff --git a/include/tscore/ink_config.h.cmake.in b/include/tscore/ink_config.h.cmake.in
index fe85c656c..f8b0c4dad 100644
--- a/include/tscore/ink_config.h.cmake.in
+++ b/include/tscore/ink_config.h.cmake.in
@@ -83,6 +83,7 @@
 #cmakedefine HAVE_BIO_GET_EX_NEW_INDEX 1
 #cmakedefine HAVE_BIO_GET_EX_DATA 1
 #cmakedefine HAVE_BIO_SET_EX_DATA 1
+#cmakedefine01 TS_DUMMY
 #cmakedefine HAVE_BIO_METH_GET_CTRL 1
 #cmakedefine HAVE_BIO_METH_GET_CREATE 1
 #cmakedefine HAVE_BIO_METH_GET_DESTROY 1
$ 
```
```
/* #undef HAVE_BIO_GET_EX_DATA */
/* #undef HAVE_BIO_SET_EX_DATA */
#define TS_DUMMY 0
```
```
#define HAVE_BIO_GET_EX_DATA 1
#define HAVE_BIO_SET_EX_DATA 1
#define TS_DUMMY 1
```

Reply:
The `#cmakedefine` directives check whether the constant evaluates as false.
https://cmake.org/cmake/help/latest/command/configure_file.html#transformations